### ISSUE TITLE
Proposal to improve the Graphite storage consistency

### ISSUE THREAD

**Issue by grzkv** [go-graphite/go-carbon#452]
# Proposal to improve the Graphite storage consistency

Note: This is just an idea, not a final solution. It is expected to be a ground for discussion and evolve into something ready for production.

## Problem

It is common to have several replicas of Graphite Whisper storage. The point of replication is to store several copies of data so that it is preserved when one of the replicas is not available. This also concerns the case when a replica is unavailable for writing. Thus, it is expected to have missing points in some replicas while they are present in others.

This situation is further complicated after points were rolled-up with time. This case produces inconsistency in data.

This is an idea to improve the consistency of storage by repairing data present on other replicas.

## Model system

These are the features of the model system that we will discuss:

1. Data is stored in 3 replicas, named `A`, `B`, and `C`.
2. Metrics are stored in files containing single retention.
3. We only consider the scenario of missing data points.
4. No roll-ups for simplicity. The solution can be extended to include them later.

## Idea for the solution

![consistent_storage](https://user-images.githubusercontent.com/3339062/156562977-17e29390-904d-4a3e-8327-7d83b7297df0.svg)

The basic idea is displayed on the diagram.

### Main components

1. `scanner` continuously scans the filesystem and gathers data about each stored file. No persistent state.
2. `coordinator` is the single source of truth and coordinates the whole process. No persistent state.
3. `filler` takes care of filling the missing data points. No persistent state.

All the components can be operated and developed independently, forming a *nix-like modular system.

### Working algorithm (numbers correspond to the numbers on diagram)

1. `scanner` scans a file.
2. `scanner` gets the pair of `(hash, latest)`: hash of the data together with the timestamp of the last written point.
3. `scanner` sends the `(hash, latest)` for the `a/b/c` metric to `coordinator` via HTTP.
4. `coordinator` stores a tree of all metrics. It writes the information to the appropriate node in that tree. Tree node consists of
   * Metric path `a/b/c`
   * Hashes of the replicas `hash<replica>`
   * Latest times of the replicas `latest<replica>`
   * The time `nextCorrectionTime` to perform the next correction if necessary.
5. `coordinator` performs a check:
   1. are there all three hashes available?
   2. if yes, is `hashA` different from `hashB` and `hashC`? (not considering the case of all three being different for now)
   3. if yes, are the `latest` times the same?
   4. if yes, is the `nextCorrectionTime` due?
   5. if yes, proceed to the next step
6. `coordinator` sends an HTTP request to `corrector` at the `A` replica that differs from `B` and `C`. The `nextCorrectionTime` is increased to prevent multiple corrections from happening at the same time of the same file.
7. `corrector` obtains data from other replicas. This can be done by talking to other `corrector`s or by reading the files.
8. `corrector` fills the missing data points in `a/b/c` on the `A` replica.
9. `corrector` sends a confirmation to `coordinator` together with the new `(hash, latest)` pair.

### Requirements for the `hash` function

* It is fast to calculate.
* It is incremental: the addition of a data point to a metric does not require full recalculation, just an O(1) update.
* It represents well the integrity of the data.

### Possible choices of the `hash` function

1. Total metric data size.
2. Number of records.
3. Any stream-optimized hash.

### `coordinator` goes down

Effect: consistency correction stops.
Fix: `coordinator` restarts, correction starts from scratch.

### `scanner` goes down

Effect: `coordinator` does not ever get all 3 `(hash, latest)` pairs; consistency correction stops.
Fix: reviving the `scanner`; correction starts again.

### `corrector` goes down during the correction procedure

If implemented by using an atomic (as much as possible) data swap.

Effect: correction procedure fails.
Fix: revive the `corrector`; after the `nextCorrectionTime` passes, it will get the same request from `coordinator`.

## Discussion

1. **No high availability for the coordinator.** It is not necessary since having it down has no significant impact, and it does not need to retain its state via WALs or anything.
2. **Using consistent key-value store** like Redis seems unjustified. Why? We do not need high availability from the `coordinator` and need additional logic.
3. **Hash functions** have their plusses and minuses. We may consider calculating them in `go-carbon` but this proposal suggests using the `scanner` to keep it simple and modular.
4. **Correction path** may go through writing to `go-carbon` input by `corrector` instead of modifying the file.
5. We may encounter scenarios when replicas are frequently **out of sync** in time, i.e. `latest` values being different. This can be solved by ignoring the freshest points.
6. One possible **optimization** for this system could be using a [Merkle-tree-like](https://en.wikipedia.org/wiki/Merkle_tree) data structure. See also [here](https://github.com/go-graphite/go-carbon/issues/452#issuecomment-1058408995).

**Comment by deniszh** [comment]
@grzkv : 
> scanner gets the pair of (hash, latest): hash of the data together with the timestamp of the last written point.
> One possible optimization for this system could be using a [Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree).

Whole process looks similar to e.g. Cassandra repairs, but SSTables in Cassandra case are immutable, and whisper files are vice versa - constantly changing. I'm afraid hash will be constantly changing and `coordinator` will be forced to "fix" all existing files that way or another, and I'm afraid Merkle tree can't help here much.
Maybe we can invent some tricky hash, which will compare separate whisper archives w/o most recent points, i.e. what you proposing in p.5

**Comment by deniszh** [comment]
>Possible choices of the hash function
>Total metric data size.
>Number of records.

Both will help with NULLs but not with wrong aggregation.

> Any stream-optimized hash.

I thinks it's only option, I'm afraid.

**Comment by deniszh** [comment]
> coordinator goes down
> Effect: consistency correction stops.
> Fix: coordinator restarts, correction starts from scratch.

correction process can be long running process and loosing all progress because of restart could be not really nice. Again, in case of Cassandra repair e.g. [Cassandra Reaper](http://cassandra-reaper.io/) using DB for persistence.

**Comment by flucrezia** [comment]
Hi,
thanks for the proposal.
I try to give my two cent.
Maybe a different approach could be a randomised algorithm; in terms of scanning time, past time window and metrics to compare. Randomisation should help in efficiency without loosing consistency.
The hash shall be computed on the output of a carbonapi-like query on past data, comparing local go-carbon data with a remote replica. If the hash is different, other replicas shall be involved and the majority win. The one not in synch then shall apply a correction routine.

**Comment by deniszh** [comment]
@flucrezia : carbonapi is a separate component, not connected to go-carbon. We can, for example, track which whisper files is reading on particular go-carbon instance and instruct coordinator to repair them first - but for that you need some api to get this info which coordinator will call.

**Comment by grzkv** [comment]
@deniszh Thanks for the feedback!

> I'm afraid hash will be constantly changing and coordinator will be forced to "fix" all existing files that way or another, 

Not sure if I explained this well enough, but the repair happens when:

1. `hash`es are different
2. `latest` times are the same

It is expected to have the hashes constantly changing. That's fine. Reparation happens only when they differ, not when they change.

Hope I understood you correctly.

> and I'm afraid Merkle tree can't help here much.

In its default form, surely not. What I meant there was to use something like a Merkle tree, but my wording does not reflect that. I've corrected the text to reflect that.

For example, we could use three Merkle trees for three replicas, and in addition to storing a hash in each node, we can store the `min(latest) over children`, for example. Then we can compare them level-by level. This is a raw idea as well, but I think there's a chance to make something like this work.

In any case, this is just an optimisation, we can probably live without it.

What do you think?

**Comment by grzkv** [comment]
@deniszh 

> Both will help with NULLs but not with wrong aggregation.

Absolutely. See model points 3 and 4.

Thing is, if we can repair data before roll-ups, they will never be a problem.

> I thinks it's only option, I'm afraid.

If we want not only NULLs but also different values it is the only reliable option. But that's not a deal-breaker at all as we can cache the state between re-calculations, and re-calculations are O(1). Also, if we use the size of the data block, we _may_ partially cover this, but that will depend on whisper internals.

**Comment by grzkv** [comment]
@deniszh 

> correction process can be long running process and loosing all progress because of restart could be not really nice. Again, in case of Cassandra repair e.g. [Cassandra Reaper](http://cassandra-reaper.io/) using DB for persistence.

Not sure we're on the same page here. We're talking about a single repair on a single whisper file. Loosing it does not seem to be a major problem for me. 

If my previous conjecture is wrong, there are still more arguments:

1. It is an exceptional situation and should not happen frequently.
2. We may cache the state to local disc and pick-up after restart. But, honestly, I don't think that would be necessary.

**Comment by grzkv** [comment]
@flucrezia 

> Maybe a different approach could be a randomised algorithm; in terms of scanning time, past time window and metrics to compare. Randomisation should help in efficiency without loosing consistency.

I agree that some randomisation may be useful. For example, when picking which files to scan first.

If I understand you correctly, you suggest not to check all data in succession, but only randomised parts. I'm not sure this would result in any significant gains as _all_ data will have to be eventually checked. Checking part of it simply leaves space for problems in unchecked parts. It will also make the whole process much less predictable and average time to full coverage longer: e.g. if we scan 50% of data every second, it will require 2 seconds to cover it progressively, but 4 seconds to cover 0.5 + 0.25 + 0.125 + 0.0625 = 94.25% of data on average.

**Comment by flucrezia** [comment]
@deniszh I didn't mean to use carbonapi component, I know it is separate, I use it with go-carbon in production everyday :) Thanks b.t.w. for the amazing work for the community. 
I meant to extract the code as a library as they do in [Moira](https://moira.readthedocs.io/en/latest/overview.html#key-features) and exploit it to query a randomly selected peer/remote replica either from the _coordinator_ or even without it in a decentralised fashion, for example, the _scanner_ could do the job.

@grzkv I agree with you as long as the assumption that we need to check all the data holds. But I don't think that is the case because a mismatch between replicas is likely to last much more than a single unit time and it is enough to detect it in at least one timestep, not in every timesteps. 
So, for example (+ correct data, - corrupted):

data: +++++---------------++++++++++++  
time: 0123456789..............n

It should be fine to detect the anomaly at whatever timestep between 5 and n.
Then we could check all other metrics of the same corrupted instance since it is likely that all metrics are corrupted and not only the one randomly picked.

Alternatively we could compare hash data over time windows larger than the interval between comparison. For example query past 5 minutes every 60 seconds so in this case we should cover all timesteps and limit the randomisation to the metric and peer selection.  

I'm saying this because, maybe I'm totally wrong, but computing an hash for each metric at each timestep and send them to the _coordinator_ seems to be a bit over-calculation. Maybe we can find a middle way.

**Comment by grzkv** [comment]
@flucrezia 

> I agree with you as long as the assumption that we need to check all the data holds. But I don't think that is the case because a mismatch between replicas is likely to last much more than a single unit time and it is enough to detect it in at least one timestep, not in every timesteps.

Agree, makes sense to me. However, it is a good question how to implement that well.

> I'm saying this because, maybe I'm totally wrong, but computing an hash for each metric at each timestep and send them to the coordinator seems to be a bit over-calculation. Maybe we can find a middle way.

This is not an intended behavior to send it for each point. The idea is to check each metric at a regular interval, and send update to the `coordinator`. Calculation itself may not be expensive at all.
